src/worker.js

```diff
@@ -0,0 +1,79 @@
+// Generated by CoffeeScript 1.12.7
+var createDb, db, sqlModuleReady;
+
+if (typeof importScripts === 'function') {
+  db = null;
+  createDb = function(data) {
+    if (db != null) {
+      db.close();
+    }
+    return db = new SQL.Database(data);
+  };
+  sqlModuleReady = initSqlJs();
+  self.onmessage = function(event) {
+    return sqlModuleReady.then(function() {
+      var buff, callback, data, done, err, result;
+      data = event['data'];
+      switch (data != null ? data['action'] : void 0) {
+        case 'open':
+          buff = data['buffer'];
+          createDb((buff ? new Uint8Array(buff) : void 0));
+          return postMessage({
+            'id': data['id'],
+            'ready': true
+          });
+        case 'exec':
+          if (db === null) {
+            createDb();
+          }
+          if (!data['sql']) {
+            throw 'exec: Missing query string';
+          }
+          return postMessage({
+            'id': data['id'],
+            'results': db.exec(data['sql'])
+          });
+        case 'each':
+          if (db === null) {
+            createDb();
+          }
+          callback = function(row) {
+            return postMessage({
+              'id': data['id'],
+              'row': row,
+              'finished': false
+            });
+          };
+          done = function() {
+            return postMessage({
+              'id': data['id'],
+              'finished': true
+            });
+          };
+          return db.each(data['sql'], data['params'], callback, done);
+        case 'export':
+          buff = db["export"]();
+          result = {
+            'id': data['id'],
+            'buffer': buff
+          };
+          try {
+            return postMessage(result, [result]);
+          } catch (error) {
+            err = error;
+            return postMessage(result);
+          }
+          break;
+        case 'close':
+          return db != null ? db.close() : void 0;
+        default:
+          throw new Error('Invalid action : ' + (data != null ? data['action'] : void 0));
+      }
+    })["catch"](function(err) {
+      return postMessage({
+        'id': event['data']['id'],
+        'error': err['message']
+      });
+    });
```

Comment:
Maybe this can be simplified to avoid chaining here ?

Reply:
Something like 

```js
sqlModuleReady
    .then(_ => handleEvent(event))
    .catch(handleError)
```

Reply:
good idea, but can we postpone refactoring src/worker.js in a separate pull-request because:
1. it seems we don't have regression testing in-place for this file, and i'm too lazy to do manual regression-testing everytime this file is modified
2. postponing 78-lines of coffescript-code-smell for future refactoring is not a big show-stopper

Reply:
I finally fixed the webworker tests that had been broken for years, and cleaned up worker.js. Closing this conversation.

However, the world has changed since this webworker handler was created, and I'm wondering whether we shouldn't just delete it and tell users to use something like [comlink](https://github.com/GoogleChromeLabs/comlink#comlink) instead.

Reply:
i'm not familiar with comlink.  i currently use this piece of portable-code to improve worker usability, which we could possibly merge into src/api.js in separate pull-request?

```js
(function () {
    /*
     * we could merge something similar to this 70-lines of code to src/api.js
     * to make webworkers easier to use in browsers.
     * it exports an async-function window.sqljsWorkerExec
     * which behaves like db.exec
     *
     * i'm not familiar with nodejs workers however.
     */
    "use strict";
    // declare toplevel variables
    let callbackDict;
    let callbackId;
    let worker;
    let sqljsWorkerExec;
    // init callbackDict
    callbackDict = {};
    // init worker
    worker = new Worker(
        // need to figure out how to make
        // the worker-script-name configurable
        "worker.sql-wasm-debug.js"
    );
    // init sqljs_worker evt-handling
    worker.addEventListener("message", function (msg) {
    /*
     * this function will handle <msg> from sqljs-worker
     */
        let callback;
        callback = callbackDict[msg.data.id];
        delete callbackDict[msg.data.id];
        if (callback) {
            callback(msg.data);
        }
    });
    sqljsWorkerExec = function (msg) {
    /*
     * this function will post <msg>.sql to sqljs-worker to exec in sqlite3
     */
        let callback;
        let err;
        // preserve stack-trace
        err = new Error();
        // normalize <msg>
        if (typeof msg === "string") {
            msg = {
                sql: msg
            };
        }
        msg = Object.assign({
            action: "exec",
            sql: "\n"
        }, msg);
        return new Promise(function (resolve, reject) {
            callback = function (data) {
                // if <data>.error, then prepend it to <err>.stack and reject
                if (data.error) {
                    err.message = data.error;
                    err.stack = data.error + "\n" + err.stack;
                    reject(err);
                    return;
                }
                resolve(data);
            };
            // increment callbackId
            callbackId = (callbackId | 0) + 1;
            msg.id = callbackId;
            callbackDict[msg.id] = callback;
            worker.postMessage(msg);
        });
    };
    // export sqljsWorkerExec
    window.sqljsWorkerExec = sqljsWorkerExec;
}());



// offload cpu and io-intensive operations to web-workers,
// while keeping the ui-responsive
(async function () {
    "use strict";
    let result;
    result = await window.sqljsWorkerExec({
        sql: `
    DROP TABLE IF EXISTS employees;
CREATE TABLE employees( id          integer,  name    text,
                          designation text,     manager integer,
                          hired_on    date,     salary  integer,
                          commission  float,    dept    integer);

  INSERT INTO employees VALUES (1,'JOHNSON','ADMIN',6,'1990-12-17',18000,NULL,4);
  INSERT INTO employees VALUES (2,'HARDING','MANAGER',9,'1998-02-02',52000,300,3);
  INSERT INTO employees VALUES (3,'TAFT','SALES I',2,'1996-01-02',25000,500,3);
  INSERT INTO employees VALUES (4,'HOOVER','SALES I',2,'1990-04-02',27000,NULL,3);
  INSERT INTO employees VALUES (5,'LINCOLN','TECH',6,'1994-06-23',22500,1400,4);
  INSERT INTO employees VALUES (6,'GARFIELD','MANAGER',9,'1993-05-01',54000,NULL,4);
  INSERT INTO employees VALUES (7,'POLK','TECH',6,'1997-09-22',25000,NULL,4);
  INSERT INTO employees VALUES (8,'GRANT','ENGINEER',10,'1997-03-30',32000,NULL,2);
  INSERT INTO employees VALUES (9,'JACKSON','CEO',NULL,'1990-01-01',75000,NULL,4);
  INSERT INTO employees VALUES (10,'FILLMORE','MANAGER',9,'1994-08-09',56000,NULL,2);
  INSERT INTO employees VALUES (11,'ADAMS','ENGINEER',10,'1996-03-15',34000,NULL,2);
  INSERT INTO employees VALUES (12,'WASHINGTON','ADMIN',6,'1998-04-16',18000,NULL,4);
  INSERT INTO employees VALUES (13,'MONROE','ENGINEER',10,'2000-12-03',30000,NULL,2);
  INSERT INTO employees VALUES (14,'ROOSEVELT','CPA',9,'1995-10-12',35000,NULL,1);

SELECT designation,COUNT(*) AS nbr, (AVG(salary)) AS avg_salary FROM employees GROUP BY designation ORDER BY avg_salary DESC;
SELECT name,hired_on FROM employees ORDER BY hired_on;
        `
    });
    console.log(JSON.stringify(result, undefined, 4));
/*
{
    "id": 1,
    "results": [
        {
            "columns": [
                "designation",
                "nbr",
                "avg_salary"
            ],
            "values": [
                [
                    "CEO",
                    1,
                    75000
                ],
                [
                    "MANAGER",
                    3,
                    54000
                ],
                [
                    "CPA",
                    1,
                    35000
                ],
                [
                    "ENGINEER",
                    3,
                    32000
                ],
                [
                    "SALES I",
                    2,
                    26000
                ],
                [
                    "TECH",
                    2,
                    23750
                ],
                [
                    "ADMIN",
                    2,
                    18000
                ]
            ]
        },
        {
            "columns": [
                "name",
                "hired_on"
            ],
            "values": [
                [
                    "JACKSON",
                    "1990-01-01"
                ],
                [
                    "HOOVER",
                    "1990-04-02"
                ],
                [
                    "JOHNSON",
                    "1990-12-17"
                ],
                [
                    "GARFIELD",
                    "1993-05-01"
                ],
                [
                    "LINCOLN",
                    "1994-06-23"
                ],
                [
                    "FILLMORE",
                    "1994-08-09"
                ],
                [
                    "ROOSEVELT",
                    "1995-10-12"
                ],
                [
                    "TAFT",
                    "1996-01-02"
                ],
                [
                    "ADAMS",
                    "1996-03-15"
                ],
                [
                    "GRANT",
                    "1997-03-30"
                ],
                [
                    "POLK",
                    "1997-09-22"
                ],
                [
                    "HARDING",
                    "1998-02-02"
                ],
                [
                    "WASHINGTON",
                    "1998-04-16"
                ],
                [
                    "MONROE",
                    "2000-12-03"
                ]
            ]
        }
    ]
}
 */
}());
```

Reply:
also thx for fixing webworker tests!